sdk/python/lib/pulumi/runtime/rpc.py

```diff
@@ -654,3 +660,15 @@ def register_resource_package(typ: str, version: str, package):
     if existing is not None:
         raise ValueError(f"Cannot re-register package {key}. Previous registration was {existing}, new registration was {package}.")
     RESOURCE_PACKAGES[key] = package
+
+RESOURCE_MODULES: Dict[str, Any] = dict()
```

Comment:
Instead of the values being typed as `Any` and assuming they are objects with `construct` function (or `construct_provider` for the case of `RESOURCE_PACKAGES`), should these just be functions typed as `Callable`?

I was kind of wondering the same for the other languages, especially now that `IResourcePackage`/`IResourceModule` are single function interfaces (e.g. C# should these just be delegates or even `Func<>`? Same for TS and Go).

Reply:
Or perhaps we should use abstract classes in Python? Something like:

```python
from abc import ABCMeta

class ResourcePackage(metaclass=ABCMeta):
    @abstractmethod
    def construct_provider(name: str, typ: str, args: Mapping[str, Any], opts: Mapping[str, Any]) -> ProviderResource:
        ...

class ResourceModule(metaclass=ABCMeta):
    @abstractmethod
    def construct(name: str, typ: str, args: Mapping[str, Any], opts: Mapping[str, Any]) -> Resource:
        ...
```

Reply:
I could have sworn that I had added a `ResourcePackage` abstract class, but I haven't. Time to dig through some old changes.

Reply:
Done.